lib/using_restricted_array.rb

```diff
@@ -6,58 +6,121 @@
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def length(array)
-  raise NotImplementedError
+  i = 0
+  until array[i] == nil
+    i += 1
+  end
+  return i
 end
 
+puts length([1,2,3,4])
+
+#********************************************************
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def print_array(array)
-  raise NotImplementedError
+  string =  "#{array[0]}"
+  i = 1
+  until array[i] == nil
+    string += " #{array[i]}"
+    i += 1
+  end
+  puts string
 end
 
+puts print_array([4,5,6,7])
+#********************************************************
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    if array[i] == value_to_find
+      return true
+    end
+  end
+  return false
 end
 
+puts search([7,8,9,0], 4, 7)
+#********************************************************
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  (length - 1).times do |i|
+    if array[i+1] > largest
+      largest = array[i + 1]
+    end
+  end
+  return largest
 end
 
+puts find_largest([2,4,6,8], 4)
+#********************************************************
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  (length - 1).times do |i|
+    if array[i+1] < smallest
+      smallest = array[i + 1]
+    end
+  end
+  return smallest
 end
 
+puts find_smallest([2,4,6,8], 4)
+#********************************************************
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n-2)
+# Space complexity: 0(1)
 def reverse(array, length)
-  raise NotImplementedError
+  reversed_array = RestrictedArray.new(length)
+  length.times do |i|
+    reversed_array[i] = array[length - 1 - i]
+  end
+  length.times do |i|
+    array[i] = reversed_array[i]
+  end
+  return array
 end
 
+puts reverse([2,4,6,8], 4)
+#********************************************************
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(logn)
+# Space complexity: 0(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  lowest_number = 0
+  highest_number = length
```

Comment:
highest should probably be `length - 1`

```suggestion
  highest_number = length - 1
```